### DISCUSSION TITLE
Generating grains based on pdf or cdf

### DISCUSSION THREAD

**Discussion by changhaoliu-UOM** [neperfepx/neper#925] [Q&A] [answered]
I am not sure how to build model based on grain volumes and the probability of the certain volumes. Below part is the tutorial on the web. But I don't understand the code and I am not sure how to write it. Firstly, what is the kind of 'mydistrib'. And is it possible to put volumes and probabilities in the file? If the sum of the probabilities is not 1, what will happen? if I want to set volums, diameq should be replaced by size. Is there any way to speed up the generation. 

Thank you very much.

Specifying an Experimental Cell Size Distribution[](https://neper.info/doc/tutorials/morpho_statistics.html#specifying-an-experimental-cell-size-distribution)

Experimental or, more generally, numerical distributions defined from a file can be specified by using option [-morpho](https://neper.info/doc/neper_t.html#cmdoption-morpho) <property>:custom(<file_name>) (instead of analytical distributions).

Consider a cell-size (diameq) distribution defined from a file, mydistrib, that contains:

0.2 10.0
0.3 20.0
0.4  5.0
i.e., 10 cells have a diameq equal to 0.2, 20 cells have a diameq equal to 0.3, and 5 cells have a diameq equal to 0.4 (for a tessellation comprising 35 cells). The numbers on the right column are actually only considered relative to each other.

Note

In contrast to analytical distributions, numerical distributions do not need to be normalized to 1.

As for an analytical distribution, if [-morpho](https://neper.info/doc/neper_t.html#cmdoption-morpho) from_morpho is used, Neper determines the number of cells from the cell size distribution and the domain size.

The corresponding tessellation can be generated as follows:

neper -T -n from_morpho                                                        \
           -morpho "diameq:custom(mydistrib),1-sphericity:lognormal(0.145,0.03)" \
           -statcell diameq                                                      \
           -o mytess

**Comment by rquey** [comment, marked as answer]
You can find an example in https://github.com/neperfepx/neper/discussions/929.
